File: app/utils/validators.py
import re

def validate_phone(phone):
    # Indian phone number validation (10 digits, optionally starting with +91)
    pattern = r'^(?:\+91)?[6-9]\d{9}$'
    return bool(re.match(pattern, phone))

def validate_email(email):
    # Basic email validation
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return bool(re.match(pattern, email))

def validate_pincode(pincode):
    # Indian PIN code validation (6 digits)
    pattern = r'^\d{6}$'
    return bool(re.match(pattern, pincode))

def validate_password(password):
    # Password validation
    # At least 8 characters
    # Must contain at least one uppercase letter
    # Must contain at least one lowercase letter
    # Must contain at least one number
    # Must contain at least one special character
    if len(password) < 8:
        return False
    if not re.search(r'[A-Z]', password):
        return False
    if not re.search(r'[a-z]', password):
        return False
    if not re.search(r'\d', password):
        return False
    if not re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
        return False
    return True

def validate_username(username):
    # Username validation
    # 5-30 characters
    # Can contain letters, numbers, and underscores
    # Must start with a letter
    pattern = r'^[a-zA-Z][a-zA-Z0-9_]{4,29}$'
    return bool(re.match(pattern, username)) 

File: app/static/js/validators.js
// Client-side validation functions
const validators = {
    phone: (value) => {
        const pattern = /^(?:\+91)?[6-9]\d{9}$/;
        return {
            isValid: pattern.test(value),
            message: 'Please enter a valid Indian phone number'
        };
    },
    
    pincode: (value) => {
        const pattern = /^\d{6}$/;
        return {
            isValid: pattern.test(value),
            message: 'Please enter a valid 6-digit PIN code'
        };
    },
    
    email: (value) => {
        const pattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
        return {
            isValid: pattern.test(value),
            message: 'Please enter a valid email address'
        };
    },
    
    password: (value) => {
        const hasUpperCase = /[A-Z]/.test(value);
        const hasLowerCase = /[a-z]/.test(value);
        const hasNumbers = /\d/.test(value);
        const hasSpecialChar = /[!@#$%^&*(),.?":{}|<>]/.test(value);
        const isValidLength = value.length >= 8;
        
        return {
            isValid: hasUpperCase && hasLowerCase && hasNumbers && hasSpecialChar && isValidLength,
            message: 'Password must be at least 8 characters with uppercase, lowercase, number and special character'
        };
    },
    
    username: (value) => {
        const pattern = /^[a-zA-Z][a-zA-Z0-9_]{4,29}$/;
        return {
            isValid: pattern.test(value),
            message: 'Username must be 5-30 characters and start with a letter'
        };
    }
};

// Real-time validation function
function validateField(input, validatorName) {
    const field = input.closest('.form-group');
    const feedbackDiv = field.querySelector('.feedback-text');
    const value = input.value.trim();
    
    if (!value && !input.required) {
        field.classList.remove('is-invalid', 'is-valid');
        feedbackDiv.textContent = '';
        return true;
    }
    
    const validation = validators[validatorName](value);
    
    if (validation.isValid) {
        field.classList.remove('is-invalid');
        field.classList.add('is-valid');
        feedbackDiv.textContent = '';
    } else {
        field.classList.remove('is-valid');
        field.classList.add('is-invalid');
        feedbackDiv.textContent = validation.message;
    }
    
    return validation.isValid;
} 